pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,338 @@
+}
+
+// PodSandboxNetworkStatus is the status of the network for a PodSandbox.
+type PodSandboxNetworkStatus struct {
+	IP string
+}
+
+// Namespaces contains paths to the namespaces.
+type Namespaces struct {
+	// Network is the path to the network namespace.
+	Network string
+}
+
+// LinuxSandBoxStatus contains status specific to Linux sandboxes.
+type LinuxPodSandboxStatus struct {
+	// Namespaces contains paths to the sandbox's namespaces.
+	Namespaces *Namespaces
+}
+
+// PodSandboxResources contains the CPU/memory resource requirements.
+type PodSandboxResources struct {
+	// CPU resource requirement.
+	CPU api.ResourceRequirements
+	// Memory resource requirement.
+	Memory api.ResourceRequirements
+}
+
+// This is to distinguish with existing ContainerID type, which includes a
+// runtime type prefix (e.g., docker://). We may rename this later.
+type RawContainerID string
+
+// ContainerRuntime provides methods for container lifecycle operations, as
+// well as listing or inspecting existing containers.
+type ContainerRuntime interface {
+	// Create creates a container in the sandbox, and returns the ID
+	// of the created container.
+	Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxID PodSandboxID) (RawContainerID, error)
+	// Start starts a created container.
+	Start(RawContainerID ContainerID) error
+	// Stop stops a running container with a grace period (i.e., timeout).
+	Stop(RawContainerID string, timeout int) error
+	// Remove removes the container.
+	Remove(RawContainerID string) error
+	// List lists the existing containers that match the ContainerFilter.
+	// The returned list should only include containers previously created
+	// by this ContainerManager.
+	List(filter ContainerFilter) ([]Container, error)
+	// Status returns the status of the container.
+	Status(RawContainerID string) (ContainerStatus, error)
+}
+
+// ContainerCommandExecutor provides methods to run a command in the container.
+// TODO: Should we merge this with ContainerRuntime?
+type ContainerCommandExecutor interface {
+	// Exec executes a command in the container.
+	Exec(RawContainerID string, cmd []string, streamOpts StreamOptions) error
+}
+
+type ContainerConfig struct {
+	// Name of the container.
+	Name string
+	// Image to use.
+	Image ImageSpec
+	// Command to execute (i.e., entrypoint for docker)
+	Command []string
+	// Args for the Command (i.e., command for docker)
+	Args []string
+	// Current working directory of the command.
+	WorkingDir string
+	// List of environment variable to set in the container
+	Env []KeyValue
+	// Mounts specifies mounts for the container
+	Mounts []Mount
+	// Labels are key value pairs that may be used to scope and select individual resources.
+	Labels Labels
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+	// Privileged runs the container in the privileged mode.
+	Privileged bool
```

Comment:
True. I agree that less overlap is a good thing. In this case specifically, we also need to consider what "privileged" means and whether kubelet can interpret all the requirements. For example, I checked docker, besides capabilities, it does not apply the default seccomp profile, and also does not meddle with paths under /proc/ (i.e., mount it as read-only or mask it). Due to the lack of clear definition in the k8s api, and the fact that kubelet doesn't have access to some low-level configuration settings (i.e., oci) in the runtime, this may be hard to achieve(?)